### PR TITLE
Create plugin module

### PR DESCRIPTION
**Pull request by ismetkizgin**

### CODE REVIEW

File: src/plugin-traverser.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  private parsePackageJson(packageJsonPath: string): {
+    pluginModule: PluginOptions;
+  } {
+    return JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
+  }
+
+  private isPluginModule(packageJson: {
+    pluginModule: PluginOptions;
+  }): boolean {
+    return (
```
Comment (mfozmen) [inline src/plugin-traverser.ts:88]: Maybe we can apply some validation pattern here to warn user on logs that what's wrong with the plugin if we identified the module as plugin.

File: src/plugin-traverser.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      pluginModuleOptions.directories &&
+      pluginModuleOptions.directories.length > 0
+    )
+      this._directories = pluginModuleOptions.directories;
+  }
+
+  public traverseDirectories(): Array<Provider<BasePlugin>> {
+    let modules = [];
+    this._directories.forEach((parentDirectory) => {
+      if (fs.existsSync(parentDirectory)) {
```
Comment (mfozmen) [inline src/plugin-traverser.ts:27]: We can use [Promise.each](http://bluebirdjs.com/docs/api/promise.each.html) to read folders asyncronous.

**Review by mfozmen — CHANGES_REQUESTED** [review]

File: src/plugin-traverser.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      pluginModuleOptions.directories.length > 0
+    )
+      this._directories = pluginModuleOptions.directories;
+  }
+
+  public async traverseDirectories(): Promise<Array<Provider<BasePlugin>>> {
+    const modules = [];
+    await bluebirdPromise.each(this._directories, (parentDirectory) => {
+      if (fs.existsSync(parentDirectory)) {
+        fs.readdirSync(parentDirectory, { withFileTypes: true })
```
Comment (mfozmen) [inline src/plugin-traverser.ts:29]: We are still reading synchronously, checkin existence synchronously even we use promise.each. We should complete file operations asynchronously.

fs.exists is deprecated by the way, so you should find another alternative like fs.open, fs.stat.

**Review by mfozmen — CHANGES_REQUESTED** [review]

File: src/plugin-traverser.ts (hunk 3 above)
Comment (ismetkizgin) [reply src/plugin-traverser.ts:29]: We cannot perform async operation in the method due to the need for output.

**Pull request merged**